d2common/d2math/vector2.go

```diff
@@ -0,0 +1,367 @@
+func (v *Vector2) Multiply(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Mul(v.x, src.X())
+	v.y.Mul(v.y, src.Y())
+
+	return v
+}
+
+// Scale this Vector2 by the given value
+func (v *Vector2) Scale(s *big.Float) d2interface.Vector2 {
+	v.x.Sub(v.x, s)
+	v.y.Sub(v.y, s)
+
+	return v
+}
+
+// Divide this Vector2 by the given Vector2
+func (v *Vector2) Divide(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Quo(v.x, src.X())
+	v.y.Sub(v.y, src.Y())
+
+	return v
+}
+
+// Negate thex and y components of this Vector2
+func (v *Vector2) Negate() d2interface.Vector2 {
+	return v.Scale(big.NewFloat(negative1))
+}
+
+// Distance calculate the distance between this Vector2 and the given Vector2
+func (v *Vector2) Distance(src d2interface.Vector2) *big.Float {
+	dist := v.DistanceSq(src)
+
+	return dist.Sqrt(dist)
+}
+
+// DistanceSq calculate the distance suared between this Vector2 and the given
+// Vector2
+func (v *Vector2) DistanceSq(src d2interface.Vector2) *big.Float {
+	delta := src.Clone().Subtract(v)
+	deltaSq := delta.Multiply(delta)
+
+	return big.NewFloat(zero).Add(deltaSq.X(), deltaSq.Y())
+}
+
+// Length returns the length of this Vector2
+func (v *Vector2) Length() *big.Float {
+	xsq, ysq := v.LengthSq()
+
+	return xsq.Add(xsq, ysq)
+}
+
+func (v *Vector2) LengthSq() (*big.Float, *big.Float) {
+	clone := v.Clone()
+	x, y := clone.X(), clone.Y()
+
+	return x.Mul(x, x), y.Mul(y, y)
+}
+
+// SetLength sets the length of this Vector2
+func (v *Vector2) SetLength(length *big.Float) d2interface.Vector2 {
+	return v.Normalize().Scale(length)
+}
+
+// Normalize Makes the vector a unit length vector (magnitude of 1) in the same
+// direction.
+func (v *Vector2) Normalize() d2interface.Vector2 {
+	xsq, ysq := v.LengthSq()
+	length := big.NewFloat(zero).Add(xsq, ysq)
+	one := big.NewFloat(1.0)
+
+	if length.Cmp(one) > 0 {
+		length.Quo(one, length.Sqrt(length))
+
+		v.x.Mul(v.x, length)
+		v.y.Mul(v.y, length)
+	}
+
+	return v
+}
+
+// NormalizeRightHand rotate this Vector2 to its perpendicular,
+//in the positive direction.
+func (v *Vector2) NormalizeRightHand() d2interface.Vector2 {
+	x := v.x
+	v.x = v.y.Mul(v.y, big.NewFloat(negative1))
+	v.y = x
+
+	return v
+}
+
+// NormalizeLeftHand rotate this Vector2 to its perpendicular,
+//in the negative1 direction.
+func (v *Vector2) NormalizeLeftHand() d2interface.Vector2 {
+	x := v.x
+	v.x = v.y
+	v.y = x.Mul(x, big.NewFloat(negative1))
+
+	return v
+}
+
+// Calculate the dot product of this Vector2 and the given Vector2
+func (v *Vector2) Dot(src d2interface.Vector2) *big.Float {
+	c := v.Clone()
+	c.X().Mul(c.X(), src.X())
+	c.Y().Mul(c.Y(), src.Y())
+
+	return c.X().Add(c.X(), c.Y())
+}
+
+// Cross Calculate the cross product of this Vector2 and the given Vector2.
+func (v *Vector2) Cross(src d2interface.Vector2) *big.Float {
+	c := v.Clone()
+	c.X().Mul(c.X(), src.X())
+	c.Y().Mul(c.Y(), src.Y())
+
+	return c.X().Sub(c.X(), c.Y())
```

Comment:
I'm not super familiar with cross product in 2D space, but from other implementations I have seen do something like `(lhs.X*rhs.Y) - (lhs.Y*rhs.X)` (note (X * Y) and not (X * X)) but again, I'm not even sure what it means in 2D space.